### DISCUSSION TITLE
How to enforce number of fields in a record?

### DISCUSSION THREAD

**Discussion by joeveiga** [millsp/ts-toolbelt#332] [Wiki]
I was wondering if there is a way to create a version of typescript Record<K, V> type where we can enforce only a single field is present. The field key is not known at compile time and can be anything (any random string in the example), but we can't have more than one.
```ts
type SingleFieldRecord<K, V> = any; // TODO

let test: SingleFieldRecord<string, number>;

test = { 'some.field': 1 }; // OK
test = { 'another.random.field': 1, 'extra.field': 2 }; // ERROR (too many fields)
```

**Comment by millsp** [comment]
I think it would be possible by inferring the keys, then using Union utilities to make that a Tuple, then return an error message if the size of the tuple is not Equal to 1.

**Reply by joeveiga** [reply]
Thanks @millsp! If you get chance to take a crack at it that'd be greatly appreciated. I'm just getting started playing with ts types and I think this is above my current level :). This is more or less my use case:
```ts
type Operator =
  | {
      // $stringIn: SingleFieldRecord<string, string[]>;
      $stringIn: {
        [key: string]: string[];
      };
    }
  | {
      // $numericGE: SingleFieldRecord<string, number>;
      $numericGE: {
        [key: string]: number;
      };
    };
  // ...

// e.g
const op: Operator = {
  $numericGE: {
    'customer.age': 18,
    // I'd like to prevent extra fields here since this api only
    // expects one single field per operator. e.g.

    'customer.height': 6 // this shouldn't be allowed 
  }
}
```
